### ISSUE TITLE
Recommended usage following Gulp task completion doesn't enforce run order

### ISSUE THREAD

**Issue by troywarr** [BrowserSync/browser-sync#537]
Using Gulp, the recommended way to reload the browser after task completion looks like this (per #535):

```js
gulp.task('bs-reload', function () {
    browserSync.reload();
});

gulp.task('watch:icons', function() {
    return gulp.watch(paths.src + "/icons/**/*", ['jade-html', 'jade-js', 'bs-reload']);
});
```

This will trigger `'bs-reload'` when changes are made to the watched files, but because Gulp's task list runs in parallel for maximum concurrency (see [async-task-support](https://github.com/gulpjs/gulp/blob/master/docs/API.md#async-task-support)), it's not necessarily the last task to run. In the example above, if the `'jade-html'` or `'jade-js'` tasks are set up to run asynchronously (for performance's sake when used in other build steps) and they take longer to complete than `'bs-reload'` takes to reload the browser, any changes will not be reflected in the browser unless it's manually refreshed after they complete.

It's possible to use the alternate `gulp.watch(glob [, opts, cb])` signature to do something like this:

```js
var runSequence = require('run-sequence');

gulp.task('watch:icons', function() {
    return gulp.watch(paths.src + "/icons/**/*", function() {
        runSequence(['jade-html', 'jade-js'], 'bs-reload');
    });
});
```

but is there a better way that doesn't rely on an additional plugin to ensure a consistent run order?

**Comment by shakyShane** [comment]
@troywarr - thanks for the excellent questions!

You are correct - should either the `jade-html` or `jade-js` task take more than a few milliseconds to complete, which, to be honest is normally ok for a lot of gulp plugins, it would still result in a race-condition between those two tasks and BrowserSync emitting the reload event to all browsers.

The answer to your question :

> is there a better way that doesn't rely on an additional plugin 

is yes. Although not a pretty solution, the following does work; the trick is to use a separate task to indicate that x number of streams must complete before reload is called. Previously where you would have something such as

```js
gulp.task('serve', function () {

    browserSync({server: './app'});

    gulp.watch('./app/*.html', ['slow-task-1', 'slow-task-2', 'bs-reload']);
});
```

You can change to 

```js
gulp.task('html-task', ['slow1', 'slow2'], reload);

gulp.task('serve', function () {

    browserSync({server: './app'});

    gulp.watch('./app/*.html', ['html-task']);
});
```

It works because now you're saying that `html-task` should not run until both of it's dependencies are complete. This does require that both `slow1` and `slow2` return their streams however, such as the following: 

```js
gulp.task('slow1', function () {
    return gulp.src('./app/*.html')
        .pipe(slowStream());
});

gulp.task('slow2', function () {
    return gulp.src('./app/*.html')
        .pipe(slowStream());
});

gulp.task('html-task', ['slow1', 'slow2'], reload);

gulp.task('serve', function () {

    browserSync({server: './app'});

    gulp.watch('./app/*.html', ['html-task']);
});
```

So there you have it - you absolutely CAN ensure a consistent run order for tasks in gulp `3.x.x`, it's just that it takes a little trickery and knowledge about gulp's task system.

Hopefully that helps you, I actually recorded a little 7 minute hangout of me coding this up as a reference for anyone else https://www.youtube.com/watch?v=yVZ7fOhO4r4

**Comment by shakyShane** [comment]
Website updated to reflect these findings http://www.browsersync.io/docs/gulp/#gulp-reload

Recipe added https://github.com/BrowserSync/recipes/tree/master/recipes/gulp.task.sequence
